spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinWebClientSender.java

```diff
@@ -34,68 +33,27 @@
  */
 class ZipkinWebClientSender extends HttpSender {
 
-	private final String endpoint;
-
 	private final WebClient webClient;
 
 	private final Duration timeout;
 
-	ZipkinWebClientSender(String endpoint, WebClient webClient, Duration timeout) {
-		this.endpoint = endpoint;
+	ZipkinWebClientSender(Encoding encoding, Factory endpointSupplierFactory, String endpoint, WebClient webClient,
+			Duration timeout) {
+		super(encoding, endpointSupplierFactory, endpoint);
 		this.webClient = webClient;
 		this.timeout = timeout;
 	}
 
 	@Override
-	public HttpPostCall sendSpans(byte[] batchedEncodedSpans) {
-		return new WebClientHttpPostCall(this.endpoint, batchedEncodedSpans, this.webClient, this.timeout);
-	}
-
-	private static class WebClientHttpPostCall extends HttpPostCall {
-
-		private final String endpoint;
-
-		private final WebClient webClient;
-
-		private final Duration timeout;
-
-		WebClientHttpPostCall(String endpoint, byte[] body, WebClient webClient, Duration timeout) {
-			super(body);
-			this.endpoint = endpoint;
-			this.webClient = webClient;
-			this.timeout = timeout;
-		}
-
-		@Override
-		public Call<Void> clone() {
-			return new WebClientHttpPostCall(this.endpoint, getUncompressedBody(), this.webClient, this.timeout);
-		}
-
-		@Override
-		protected Void doExecute() {
-			sendRequest().block();
-			return null;
-		}
-
-		@Override
-		protected void doEnqueue(Callback<Void> callback) {
-			sendRequest().subscribe((entity) -> callback.onSuccess(null), callback::onError);
-		}
-
-		private Mono<ResponseEntity<Void>> sendRequest() {
-			return this.webClient.post()
-				.uri(this.endpoint)
-				.headers(this::addDefaultHeaders)
-				.bodyValue(getBody())
-				.retrieve()
-				.toBodilessEntity()
-				.timeout(this.timeout);
-		}
-
-		private void addDefaultHeaders(HttpHeaders headers) {
-			headers.addAll(getDefaultHeaders());
-		}
-
+	void postSpans(URI endpoint, HttpHeaders headers, byte[] body) {
+		this.webClient.post()
+			.uri(endpoint)
+			.headers((h) -> h.addAll(headers))
+			.bodyValue(body)
+			.retrieve()
+			.toBodilessEntity()
+			.timeout(this.timeout)
+			.block();
```

Comment:
I guess the contract of `BytesMessageSender` is that `send` blocks, right?

This `.block()` on the webclient will block one event loop thread. Maybe it would be better to schedule that on a threadpool?

Reply:
We have this because the `RestTemplate` is in `spring-web`, which is not on the classpath when using webflux. That's why we added the webflux one.

Maybe we could implement one based on the JDK httpclient, then we can get rid of the webflux one which is misusing the event loop with the `.block` here.

Reply:
I've opened #39545 for that.

Reply:
We still have to decide what to do with the webclient one. Calling `.block` is bad, removing the webclient based sender will break users who currently only have webflux on the classpath. Maybe using an executor where we run the webclient on?

What's your take on this, @wilkinsona ?

Reply:
We already call `block()` in some circumstances:

https://github.com/spring-projects/spring-boot/blob/2abc25d51cbb672ef25bdb8141bb5fef25088404/spring-boot-project/spring-boot-actuator-autoconfigure/src/main/java/org/springframework/boot/actuate/autoconfigure/tracing/zipkin/ZipkinWebClientSender.java#L76

That means that it's not clear to me how much worse this change is making things. That said, dropping the `WebClient`-based sender in favour of one that uses `HttpClient` makes sense. I think we could accept this change as proposed and then #39545 would make the use of `block()` irrelevant anyway.

Reply:
Thanks. Sounds good.

Reply:
I agree lets leave it alone. the part of the code in question just moved from one place to another. The "which client" thing is the same concern as before and can be done in a follow-up (with folks who know better about pros/cons and ways out than me!)